perceiver_pytorch/encoders.py

```diff
@@ -0,0 +1,353 @@
+import torch
+import torch.nn.functional as F
+import numpy as np
+import math
+import einops
+
+def extract_image_patches(x, kernel, stride=1, dilation=1):
+    # Do TF 'SAME' Padding
+    b, c, h, w = x.shape
+    h2 = math.ceil(h / stride)
+    w2 = math.ceil(w / stride)
+    pad_row = (h2 - 1) * stride + (kernel - 1) * dilation + 1 - h
+    pad_col = (w2 - 1) * stride + (kernel - 1) * dilation + 1 - w
+    x = F.pad(x, (pad_row // 2, pad_row - pad_row // 2, pad_col // 2, pad_col - pad_col // 2))
+
+    # Extract patches
+    patches = x.unfold(2, kernel, stride).unfold(3, kernel, stride)
+    patches = patches.permute(0, 4, 5, 1, 2, 3).contiguous()
+
+    return patches.view(b, -1, patches.shape[-2], patches.shape[-1])
+
+
+def reverse_space_to_depth(
+        frames: np.ndarray, temporal_block_size: int = 1, spatial_block_size: int = 1
+) -> np.ndarray:
+    """Reverse space to depth transform."""
+    if len(frames.shape) == 4:
```

Comment:
Could just add a comment, saying dim=4 is an image, dim=5 is a video

Reply:
Done